Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp

```diff
@@ -412,20 +412,143 @@ void RealtimeVideoCaptureSource::setSizeFrameRateAndZoom(std::optional<int> widt
+}
+
+auto RealtimeVideoCaptureSource::takePhoto(PhotoSettings&& photoSettings) -> Ref<TakePhotoNativePromise>
+{
+    ASSERT(isMainThread());
+
+    TakePhotoNativePromise::Producer takePhotoProducer;
+    Ref<TakePhotoNativePromise> takePhotoPromise = takePhotoProducer;
+
+    queueAndProcessSerialAction([photoSettings = WTFMove(photoSettings), protectedThis = Ref { *this }, takePhotoProducer = WTFMove(takePhotoProducer), this]() mutable -> Ref<GenericPromise> {
+
+        if (isEnded()) {
+            takePhotoProducer.reject("Track has ended"_s);
+            return GenericPromise::createAndResolve();
+        }
+
+        if ((photoSettings.imageHeight && !photoSettings.imageWidth) || (!photoSettings.imageHeight && photoSettings.imageWidth)) {
+            IntSize sanitizedSize;
+            if (photoSettings.imageHeight)
+                sanitizedSize.setHeight(*photoSettings.imageHeight);
+            if (photoSettings.imageWidth)
+                sanitizedSize.setWidth(*photoSettings.imageWidth);
+
+            auto intrinsicSize = this->intrinsicSize();
+            if (!sanitizedSize.height())
+                sanitizedSize.setHeight(sanitizedSize.width() * (intrinsicSize.height() / static_cast<double>(intrinsicSize.width())));
+            else if (!sanitizedSize.width())
+                sanitizedSize.setWidth(sanitizedSize.height() * (intrinsicSize.width() / static_cast<double>(intrinsicSize.height())));
+
+            photoSettings.imageHeight = sanitizedSize.height();
+            photoSettings.imageWidth = sanitizedSize.width();
+        }
+
+        std::optional<CaptureSizeFrameRateAndZoom> newPresetForPhoto;
+        if (photoSettings.imageHeight || photoSettings.imageWidth) {
+            newPresetForPhoto = bestSupportedSizeFrameRateAndZoomConsideringObservers(photoSettings.imageWidth, photoSettings.imageHeight, { }, { });
+            ERROR_LOG_IF(loggerPtr() && !newPresetForPhoto, LOGIDENTIFIER, "unable to find a preset to match the size of requested photo, using current preset");
+
+            if (newPresetForPhoto && m_currentPreset && m_currentPreset->size() == newPresetForPhoto->encodingPreset->size())
+                newPresetForPhoto = { };
+        }
+
+        std::optional<CaptureSizeFrameRateAndZoom> configurationToRestore;
+        if (newPresetForPhoto) {
+            configurationToRestore = {
+                { m_currentPreset },
+                size(),
+                frameRate(),
+                zoom()
+            };
+
+            // 3.2.2 - Devices MAY temporarily stop streaming data, reconfigure themselves with the appropriate photo
+            // settings, take the photo, and then resume streaming. In this case, the stopping and restarting of
+            // streaming SHOULD cause onmute and onunmute events to fire on the track in question.
+            if (!muted()) {
+                setMuted(true);
+                m_mutedForPhotoCapture = true;
```

Comment:
One potential issue is that this will make our UI button flaky for a bit, while we are not stopping capture.
I do not have a good suggestion though, I am tempted to leave this signalling aside for now.